worlds/marioland2/logic.py

```diff
@@ -0,0 +1,607 @@
+            reachable_coins += 8
+    return coins <= reachable_coins
+
+
+def tree_zone_3_normal_exit(state, player):
+    return not is_auto_scroll(state, player, "Tree Zone 3")
+
+
+def tree_zone_3_coins(state, player, coins):
+    if is_auto_scroll(state, player, "Tree Zone 3"):
+        return coins <= 4
+    if coins <= 19:
+        return True
+    elif state.has_any(["Mushroom", "Fire Flower"], player) and coins <= 21:
+        return True
+    return state.has("Carrot", player)
+
+
+def tree_zone_4_normal_exit(state, player):
+    return has_pipe_down(state, player) and tree_zone_4_midway_bell(state, player)
+
+
+def tree_zone_4_midway_bell(state, player):
+    return ((has_pipe_right(state, player) and has_pipe_up(state, player))
+            or state.has("Tree Zone 4 Midway Bell", player))
+
+
+def tree_zone_4_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Tree Zone 4")
+    reachable_coins = 0
+    if has_pipe_up(state, player):
+        reachable_coins += 14
+        if has_pipe_right(state, player):
+            reachable_coins += 4
+            if has_pipe_down(state, player):
+                reachable_coins += 10
+                if not auto_scroll:
+                    reachable_coins += 46
+    elif state.has("Tree Zone 4 Midway Bell", player):
+        if not auto_scroll:
+            if has_pipe_left(state, player):
+                reachable_coins += 18
+                if has_pipe_down(state, player):
+                    reachable_coins += 10
+                    if has_pipe_up(state, player):
+                        reachable_coins += 46
+        elif has_pipe_down(state, player):
+            reachable_coins += 10
+    return coins <= reachable_coins
+
+
+def tree_zone_5_boss(state, player):
+    return has_pipe_right(state, player) and (has_pipe_up(state, player) or state.has("Carrot", player))
+
+
+def tree_zone_5_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Tree Zone 5")
+    reachable_coins = 0
+    # Not actually sure if these platforms can be randomized / can make the coin blocks unreachable from below
+    if ((not state.multiworld.worlds[player].options.randomize_platforms)
+            or state.has_any(["Mushroom", "Fire Flower"], player)):
+        reachable_coins += 2
+    if state.has_any(["Mushroom", "Fire Flower"], player):
+        reachable_coins += 2
+    if state.has("Carrot", player):
+        reachable_coins += 18
+        if has_pipe_up(state, player) and not auto_scroll:
+            reachable_coins += 13
+    elif has_pipe_up(state, player):
+        reachable_coins += 13
+    return coins <= reachable_coins
+
+
+def pumpkin_zone_1_normal_exit(state, player):
+    return pumpkin_zone_1_midway_bell(state, player)
+
+
+def pumpkin_zone_1_midway_bell(state, player):
+    return ((has_pipe_down(state, player) and not is_auto_scroll(state, player, "Pumpkin Zone 1"))
+            or state.has("Pumpkin Zone 1 Midway Bell", player))
```

Comment:
It seems like this can always be false, specifically if Midway Bells are not shuffled and the level is an autoscroller. This leads to "Pumpkin Zone 1 - Normal Exit" being inaccessible.

Reply:
https://github.com/ArchipelagoMW/Archipelago/pull/2730/commits/d1f136b9ead4a9a32cd396a7b59474859bcb8d84